src/Simulation/NativeSparseSimulator/SparseSimulator.h

```diff
@@ -148,7 +170,9 @@ class SparseSimulator
 			_execute_queued_ops(qubit_id);
 			
 			if (!_quantum_state->is_qubit_zero(qubit_id)){
-				throw std::runtime_error("Released qubit not in zero state");
+				_quantum_state->Reset(qubit_id);
+				_set_qubit_to_zero(qubit_id);
+				return false;
```

Comment:
I appreciate the callout! This is indeed an area fraught with options, and one we've gone back and forth on. We fixed the fullstate simulator's QIR execution to treat releasing non-zero qubits as errors, making it consistent with how local simulation via C# works, only to get the feedback that this behavior interferes with running QIR programs from non-Q# languages where this convention is not enforced. Given that feedback, I think there was desire to remove the error even for the fullstate simulator, at least when running QIR, but I believe it's still something of an open question.

Given all of that, I think the behavior written here makes sense, even though it can cause side effects. Changing it to be an error is something we should leave until we have a more definitive decision regarding the behavior of our other simulators.

Reply:
Sounds good. Whatever way we decide to go on that, we should probably be consistent and make sure that's well documented. I think there's likely still a lot of parts of the Q# docs, for instance, that assume that implicit resets don't happen after releasing qubits, such that optimizations from coherent unpreparation are allowed — if we decide for always implicitly resetting, those docs and samples will likely need to be updated.

Reply:
https://github.com/microsoft/qsharp-compiler/issues/1315
Marking as resolved.